### PR TITLE
Make sure a report doesn't get deleted upon the deletion of its video

### PR DESCRIPTION
**Pull request by rigelk**
Also contained in this PR:
- follows/moderation table empty states
- moderation tables generic filter
- moderation comment indicator
- moderation tables column space optimizations and ordering
- specific to video-abuse-list:
  - deletion indicator
  - additional quick actions for video and reporter: mute, blacklist, delete
  - video thumbnail and channel displayed along video title
  - reports stats for reporter and reportee
  - position of report among reports for a given video

Screenshot_2020-04-18 Video abuses list - PeerTube

### CODE REVIEW

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.ts [hunk 1]
```diff
@@ -119,6 +218,11 @@ export class VideoAbuseListComponent extends RestTable implements OnInit {
     return Account.CREATE_BY_STRING(account.name, account.host)
   }
 
+  setTableFilter (filter: string) {
+    const filterInput = document.getElementById('table-filter') as HTMLInputElement
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.ts:222]: Why don't you use `ViewChild` to get the filter input?

File: client/src/app/header/search-typeahead.component.ts [hunk 2]
```diff
@@ -146,11 +146,15 @@ export class SearchTypeaheadComponent implements OnInit, OnDestroy {
     }
   }
 
+  isOnSearch () {
+    return window.location.pathname === '/search'
```
Comment (Chocobozzz) [inline client/src/app/header/search-typeahead.component.ts:150]: :+1:

File: client/src/app/shared/rest/rest-table.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export abstract class RestTable {
   abstract sort: SortMeta
   abstract pagination: RestPagination
 
-  protected search: string
+  search: string
```
Comment (Chocobozzz) [inline client/src/app/shared/rest/rest-table.ts:14]: Where is it used outside the children?

File: server/helpers/middlewares/video-abuses.ts [hunk 4]
```diff
@@ -1,9 +1,17 @@
 import { Response } from 'express'
 import { VideoAbuseModel } from '../../models/video/video-abuse'
+import { fetchVideo } from '../video'
 
-async function doesVideoAbuseExist (abuseIdArg: number | string, videoId: number, res: Response) {
+async function doesVideoAbuseExist (abuseIdArg: number | string, videoUUID: string, res: Response) {
```
Comment (Chocobozzz) [inline server/helpers/middlewares/video-abuses.ts:5]: It breaks the API no?

File: server/helpers/middlewares/video-abuses.ts [hunk 5]
```diff
@@ -1,9 +1,17 @@
...
   const abuseId = parseInt(abuseIdArg + '', 10)
-  const videoAbuse = await VideoAbuseModel.loadByIdAndVideoId(abuseId, videoId)
+  let videoAbuse = await VideoAbuseModel.loadByIdAndVideoId(abuseId, null, videoUUID)
+
+  if (!videoAbuse) {
+    const userId = res.locals.oauth ? res.locals.oauth.token.User.id : undefined
```
Comment (Chocobozzz) [inline server/helpers/middlewares/video-abuses.ts:10]: `res.locals.oauth?.token.User.id` should work now

File: server/initializers/migrations/0490-abuse-video.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  queryInterface: Sequelize.QueryInterface
+  sequelize: Sequelize.Sequelize
+}): Promise<void> {
+
+  const deletedVideo = {
+    type: Sequelize.JSONB,
+    allowNull: true
+  }
+  await utils.queryInterface.addColumn('videoAbuse', 'deletedVideo', deletedVideo)
+  await utils.sequelize.query(`ALTER TABLE "videoAbsue" ALTER COLUMN "videoId" DROP NOT NULL;`)
```
Comment (Chocobozzz) [inline server/initializers/migrations/0490-abuse-video.ts:14]: typo

File: server/initializers/migrations/0490-abuse-video.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  await utils.sequelize.query(`ALTER TABLE "videoAbuse" DROP CONSTRAINT IF EXISTS "videoAbuse_videoId_fkey";`)
+  await utils.sequelize.query(`ALTER TABLE "videoAbuse" ADD CONSTRAINT "videoAbuse_videoId_fkey"
```
Comment (Chocobozzz) [inline server/initializers/migrations/0490-abuse-video.ts:16]: I don't think you need to add the constraint: sequelize will do it for you on startup

File: server/models/server/server-blocklist.ts [hunk 8]
```diff
@@ -120,16 +120,27 @@ export class ServerBlocklistModel extends Model<ServerBlocklistModel> {
...
       limit: count,
       order: getSort(sort),
       where: {
-        accountId
+        accountId,
+        ...searchAttribute(search, '$BlockedServer.host$')
       }
     }
 
+    console.log(search)
```
Comment (Chocobozzz) [inline server/models/server/server-blocklist.ts:142]: o/

File: server/models/utils.ts [hunk 9]
```diff
@@ -207,6 +207,16 @@ function buildDirectionAndField (value: string) {
   return { direction, field }
 }
 
+function searchAttribute (sourceField, targetField) {
+  return sourceField
```
Comment (Chocobozzz) [inline server/models/utils.ts:211]:
```
if (sourceField) {
  return ...
}

return ...
```

File: server/models/video/video-abuse.ts [hunk 10]
```diff
@@ -7,14 +9,164 @@ import {
...
+            ),
+            'countReportsForReportee'
+          ]
+        ]
+      },
+      include: [
+        {
+          model: AccountModel,
+          required: true,
+          where: { ...searchAttribute(options.searchReporter, 'name') }
```
Comment (Chocobozzz) [inline server/models/video/video-abuse.ts:145]: `where: searchAttribute(options.searchReporter, 'name')`

File: server/models/account/account-blocklist.ts [hunk 11]
```diff
@@ -111,16 +111,36 @@ export class AccountBlocklistModel extends Model<AccountBlocklistModel> {
...
+      order: getSort(sort)
+    }
+
+    const where = {
+      accountId
     }
 
+    if (search) {
+      Object.assign(where, {
+        [Op.or]: [
```
Comment (Chocobozzz) [inline server/models/account/account-blocklist.ts:135]:
```
        [Op.or]: [
          searchAttribute(search, '$BlockedAccount.name$'),
          searchAttribute(search, '$BlockedAccount.Actor.url$')
        ]
```

File: server/models/video/video-abuse.ts [hunk 12]
```diff
@@ -7,14 +9,164 @@ import {
...
+}
+
+@Scopes(() => ({
+  [ScopeNames.FOR_API]: (options: {
+    search?: string
+    searchReporter?: string
+    searchVideo?: string
+    searchVideoChannel?: string
+    serverAccountId: number
+    userAccountId: any
```
Comment (Chocobozzz) [inline server/models/video/video-abuse.ts:34]: `any`?

File: server/models/video/video-abuse.ts [hunk 13]
```diff
@@ -7,14 +9,164 @@ import {
...
+      }
+    }
+
+    if (options.search) {
+      where = Object.assign(where, {
+        [Op.or]: [
+          {
+            [Op.and]: [
+              { videoId: { [Op.not]: null } },
+              { '$Video.name$': { [Op.iLike]: `%${options.search}%` } }
```
Comment (Chocobozzz) [inline server/models/video/video-abuse.ts:48]: You don't use `searchAttribute`?

File: server/models/video/video-abuse.ts [hunk 14]
```diff
@@ -7,14 +9,164 @@ import {
...
+      })
+    }
 
+    return {
+      attributes: {
+        include: [
+          [
+            literal(
+              '(' +
+                'SELECT t.count ' +
```
Comment (Chocobozzz) [inline server/models/video/video-abuse.ts:80]: `SELECT COUNT(*) FROM "videoAbuse" WHERE "videoId" = "VideoAbuseModel"."videoId"`?

File: server/models/video/video-abuse.ts [hunk 15]
```diff
@@ -7,14 +9,164 @@ import {
...
+                'INNER JOIN "account" ON "videoChannel"."accountId" = "account"."id" ' +
+                'WHERE "account"."id" = "VideoAbuseModel"."reporterAccountId" ' +
+              ')'
+            ),
+            'countReportsForReporter'
+          ],
+          [
+            literal(
+              '(' +
+                'WITH ' +
```
Comment (Chocobozzz) [inline server/models/video/video-abuse.ts:121]:
```
SELECT COUNT(*) FROM "videoAbuse" 
INNER JOIN "video" ON "video"."id" = "videoAbuse"."videoId"
INNER JOIN "videoChannel" ON "videoChannel"."id" = "video"."channelId"
INNER JOIN "account" ON "videoChannel"."accountId" = "VideoAbuseModel.Video.VideoChannel.accountId"
```

You'll have to find the real column name generated by sequelize of `"VideoAbuseModel.Video.VideoChannel.accountId"`

File: server/models/video/video.ts [hunk 16]
```diff
@@ -798,6 +798,35 @@ export class VideoModel extends Model<VideoModel> {
...
+    if (!Array.isArray(instance.VideoAbuses)) {
+      instance.VideoAbuses = await instance.$get('VideoAbuses')
+
+      if (instance.VideoAbuses.length === 0) return undefined
+    }
+
+    const details = instance.toFormattedDetailsJSON()
+
+    for (const abuse of instance.VideoAbuses) {
+      tasks.push((_ => {
```
Comment (Chocobozzz) [inline server/models/video/video.ts:816]:
```
abuse.deletedVideo = details
tasks.push(abuse.save({ transaction: options.transaction }))
```

File: server/typings/models/video/video-abuse.ts [hunk 17]
```diff
@@ -16,12 +16,12 @@ export type MVideoAbuseId = Pick<VideoAbuseModel, 'id'>
...
   MVideoAbuse &
   Pick<VideoAbuseModel, 'toActivityPubObject'> &
-  Use<'Video', MVideo>
+  Use<'Video', MVideoAccountLightBlacklistAllFiles>
 
 export type MVideoAbuseAccountVideo =
   MVideoAbuse &
   Pick<VideoAbuseModel, 'toActivityPubObject'> &
-  Use<'Video', MVideo> &
+  Use<'Video', MVideoAccountLightBlacklistAllFiles> &
```
Comment (Chocobozzz) [inline server/typings/models/video/video-abuse.ts:24]: Why did you change this definition?

File: server/typings/models/video/video-abuse.ts [hunk 18]
```diff
@@ -16,12 +16,12 @@ export type MVideoAbuseId = Pick<VideoAbuseModel, 'id'>
 export type MVideoAbuseVideo =
   MVideoAbuse &
   Pick<VideoAbuseModel, 'toActivityPubObject'> &
-  Use<'Video', MVideo>
+  Use<'Video', MVideoAccountLightBlacklistAllFiles>
```
Comment (Chocobozzz) [inline server/typings/models/video/video-abuse.ts:19]: Did you check references of this definition have these extra Video fields?

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 19]
```diff
@@ -40,47 +51,141 @@
...
+      <td *ngIf="!videoAbuse.video.deleted">
+        <a [href]="getVideoUrl(videoAbuse)" class="video-table-video-link" i18n-title title="Open video in a new tab" target="_blank" rel="noopener noreferrer">
+          <div class="video-table-video">
+            <div class="video-table-video-image">
+              <img [src]="videoAbuse.video.thumbnailPath">
+              <span
+                class="video-table-video-image-label" *ngIf="videoAbuse.count > 1"
+                i18n-title title="This video has been reported multiple times."
+              >{{ videoAbuse.nth }}/{{ videoAbuse.count }}</span>
+            </div>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:63]: Please add `\n` so it's easier to see the main blocks

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 20]
```diff
@@ -40,47 +51,141 @@
...
+            <div class="video-table-video-text">
+              <div>
+                {{ videoAbuse.video.name }}
+                <span *ngIf="!videoAbuse.video.blacklisted" class="glyphicon glyphicon-new-window"></span>
+                <span *ngIf="videoAbuse.video.blacklisted" i18n-title title="Video was blacklisted" class="glyphicon glyphicon-ban-circle"></span>
+              </div>
+              <div class="text-muted">by {{ videoAbuse.video.channel?.displayName }} on {{ videoAbuse.video.channel?.host }} </div>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:70]: miss i18n

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 21]
```diff
@@ -40,47 +51,141 @@
...
-      <td class="c-hand" [pRowToggler]="videoAbuse">
+      <td *ngIf="videoAbuse.video.deleted" class="c-hand" [pRowToggler]="videoAbuse">
+        <div class="video-table-video" i18n-title title="Video was deleted">
+          <div class="video-table-video-image"><span i18n>Deleted</span></div>
+          <div class="video-table-video-text">
+            <div>
+              {{ videoAbuse.video.name }}
+              <span class="glyphicon glyphicon-trash"></span>
+            </div>
+            <div class="text-muted">by {{ videoAbuse.video.channel?.displayName }} on {{ videoAbuse.video.channel?.host }} </div>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:84]: i18n

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 22]
```diff
@@ -40,47 +51,141 @@
...
+              <div class="text-muted">by {{ videoAbuse.video.channel?.displayName }} on {{ videoAbuse.video.channel?.host }} </div>
+            </div>
+          </div>
         </a>
       </td>
 
-      <td class="c-hand" [pRowToggler]="videoAbuse">
+      <td *ngIf="videoAbuse.video.deleted" class="c-hand" [pRowToggler]="videoAbuse">
+        <div class="video-table-video" i18n-title title="Video was deleted">
+          <div class="video-table-video-image"><span i18n>Deleted</span></div>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:78]: Don't use 1 line for multiple blocks

````
<div>
  <span>...</span>
</div>

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 23]
```diff
@@ -40,47 +51,141 @@
...
+                    >
+                    <div>
+                      <span class="text-muted">{{ createByString(videoAbuse.reporterAccount) }}</span>
+                    </div>
+                  </div>
+                  <a [routerLink]="[ '/admin/moderation/video-abuses/list' ]" [queryParams]="{ 'search': videoAbuse.reporterAccount.displayName }" class="ml-auto text-muted video-details-links" i18n>
+                    {videoAbuse.countReportsForReporter, plural, =1 {1 report} other {{{ videoAbuse.countReportsForReporter }} reports}}<span class="ml-1 glyphicon glyphicon-flag"></span>
+                  </a>
+                </span>
+              </div>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:130]: `\n`

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html [hunk 24]
```diff
@@ -40,47 +51,141 @@
...
+                <span class="col-3 moderation-expanded-label" i18n>Note</span>
                 <span class="col-9 moderation-expanded-text" [innerHTML]="videoAbuse.moderationCommentHtml"></span>
               </div>
             </div>
 
             <div class="col-4">
-              <div class="screenratio" [innerHTML]="videoAbuse.embedHtml"></div>
+              <div class="screenratio">
+                <div *ngIf="videoAbuse.video.deleted || videoAbuse.video.blacklisted">
+                  <span i18n>The video was {{ videoAbuse.video.deleted ? 'deleted' : 'blacklisted' }}</span>
```
Comment (Chocobozzz) [inline client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.html:169]: For translators, I think it would be clearer with

````
<span i18n *ngIf="videoAbuse.video.deleted">This video was deleted</span>
<span i18n *ngIf="!videoAbuse.video.deleted">This video was blacklistedspan>
````

File: client/src/app/shared/rest/rest-table.ts (hunk 3 above)
Comment (rigelk) [reply client/src/app/shared/rest/rest-table.ts:14]: It is used in the children's _templates_.

File: server/models/server/server-blocklist.ts (hunk 8 above)
Comment (rigelk) [reply server/models/server/server-blocklist.ts:142]: :sweat_smile:

File: server/models/video/video-abuse.ts (hunk 10 above)
Comment (rigelk) [reply server/models/video/video-abuse.ts:145]: ah. of course :octopus:

File: server/models/video/video-abuse.ts (hunk 14 above)
Comment (rigelk) [reply server/models/video/video-abuse.ts:80]: Yes, I got hyped by `PARTITION BY` :grimacing:

File: server/helpers/middlewares/video-abuses.ts (hunk 4 above)
Comment (rigelk) [reply server/helpers/middlewares/video-abuses.ts:5]: Yes, but I made sure the rest of the API isnt used in our code.

File: client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.ts (hunk 1 above)
Comment (rigelk) [reply client/src/app/+admin/moderation/video-abuse-list/video-abuse-list.component.ts:220]: Because the element is usually not generated (it is within p-table's rendering) before the AfterContentChecked stage.

File: server/models/video/video-abuse.ts (hunk 15 above)
Comment (rigelk) [reply server/models/video/video-abuse.ts:117]: Actually sequelize has a syntax *for associations*: "Video->VideoChannel"."accountId".

File: server/typings/models/video/video-abuse.ts (hunk 18 above)
Comment (rigelk) [reply server/typings/models/video/video-abuse.ts:19]: Indeed, they don't for `MVideoAbuseVideo`.

File: server/typings/models/video/video-abuse.ts (hunk 17 above)
Comment (rigelk) [reply server/typings/models/video/video-abuse.ts:24]: `toFormattedJSON` requires `MVideoAbuseFormattable` which is called for the listing, and required by `VideoAbuseAuditView`.
